core/CronArchive.php

```diff
@@ -894,6 +894,24 @@ private function invalidateWithSegments($idSites, $date, $period, $_forceInvalid
                 if ($this->canWeSkipInvalidatingBecauseThereIsAUsablePeriod($params, $doNotIncludeTtlInExistingArchiveCheck)) {
                     $this->logger->debug('  Found usable archive for {archive}, skipping invalidation.', ['archive' => $params]);
                 } else {
+                    if (empty($this->segmentArchiving)) {
+                        // might not be initialised if init is not called
+                        $this->segmentArchiving = StaticContainer::get(SegmentArchiving::class);
+                    }
+                    $segmentHash = $params->getSegment()->getHash();
+                    $segmentInfo = $this->segmentArchiving->findSegmentForHash($segmentHash, $idSite);
+                    $periodEnd = $params->getPeriod()->getDateEnd();
+
+                    if ($segmentInfo) {
+                        $segmentArchiveStartDate = $this->segmentArchiving->getReArchiveSegmentStartDate($segmentInfo);
+
+                        if ($segmentArchiveStartDate !== null && $segmentArchiveStartDate->isLater($periodEnd->getEndOfDay())) {
+                            // the system is not allowed to invalidate reports for this period
+                            // automatically, only a user can specifically invalidate
+                            continue;
+                        }
+                    }
```

Comment:
This code (the function calls and ifs) looks very similar to the above code. Can it be moved to a method to reduce duplication? Maybe a method in the `SegmentArchiving` class?

Reply:
@flamisz this wasn't addressed, did you discuss it w/ @tsteur or something?

Reply:
I didn't any useful new function, but refactored both files, mostly get rid off temporary variables.

Reply:
I was thinking something like:

SegmentArchiving.php
```
public function getSegmentReArchiveDateIfPastEndOfArchivingPeriod($params)
{
    $segmentInfo = $this->findSegmentForHash($params->getSegment()->getHash(), $params->getSite()->getId());

        if (!$segmentInfo) {
            return false;
        }

        $segmentArchiveStartDate = $this->getReArchiveSegmentStartDate($segmentInfo);

    if ($segmentArchiveStartDate !==null && $segmentArchiveStartDate->isLater($params->getPeriod()->getDateEnd()->getEndOfDay())) {
        return $segmentArchiveDate;
    }

    return false;
}
```

then in Loader.php:

```
    public function canSkipArchiveForSegment()
    {
        /** @var SegmentArchiving */
        $segmentArchiving = StaticContainer::get(SegmentArchiving::class);
        $segmentReArchiveDate = $segmentArchiving->getSegmentReArchiveDateIfPastEndOfArchivingPeriod($this->params);

        if (!empty($segmentReArchiveDate)) {
            $doneFlag = Rules::getDoneStringFlagFor(
                [$params->getSite()->getId()],
                $params->getSegment(),
                $params->getPeriod()->getLabel(),
                $params->getRequestedPlugin()
            );

            // if there is no invalidation where the report is null, we can skip
            // if we have invalidations for the period and name, but only for a specific reports, we can skip
            // if the report is not null we only want to rearchive if we have invalidation for that report
            // if we don't find invalidation for that report, we can skip
            return !$this->dataAccessModel->hasInvalidationForPeriodAndName($idSite, $params->getPeriod(), $doneFlag, $params->getArchiveOnlyReport());
        }

        return false;
    }
```

and above in CronArchive.php:

```
$segmentReArchiveDate = $segmentArchiving->getSegmentReArchiveDateIfPastEndOfArchivingPeriod($this->params);
if (!empty($segmentReArchiveDate)) {
    // the system is not allowed to invalidate reports for this period
    // automatically, only a user can specifically invalidate
    continue;
}
```

Maybe this doesn't work or it's not a good idea for another reason, I'll just take your word for it since it's not a huge deal.